### PR TITLE
XI-6316 Add referenceLink to question object

### PR DESCRIPTION
**Pull request by Janis4411**
This MR introduces the reviewLink to the question object. Furthermore the results table has been expanded by the Review Link column to hold the reviewLink for each question, if there is one.
Also added tests for the Result component.

Screenshot with no question having a reviewLink:

Screenshot 2024-04-16 at 11 29 43

Screenshot with at least one question having a reviewLink:

Screenshot 2024-04-17 at 09 47 17

## Checklist

_Make sure all these items are checked before submitting the pull request_

- [ ] All checks pass successfully
- [ ] All related commits are squashed together
- [ ] The code follows the project's coding standards
- [ ] The changes are properly documented / the relevant documentation is updated (if applicable)
- [ ] Appropriate labels are added to this pull request

### CODE REVIEW

**Comment by Janis4411** [comment]
I'll take care of the pipeline errors! 🙂

File: src/App.test.tsx [hunk 1]
```diff
@@ -92,8 +92,10 @@ describe('App component', () => {
       });
 
       const result = screen.queryByText(/Result/);
-
       expect(result).toBeInTheDocument();
+
+      const referenceLink = screen.getByRole('link', { name: 'Reference' });
```
Comment (christophblessing) [inline src/App.test.tsx:97]: From what I can see is that this test is obsolete because the coverage is in `Result.test.tsx`

File: src/locales/en.ts [hunk 2]
```diff
@@ -35,6 +35,8 @@ const en = {
     correct: 'Correctly answered',
     question: 'Question',
     attempts: 'Attempts',
+    reference: 'Reference',
+    referenceLink: 'Review link',
```
Comment (christophblessing) [inline src/locales/en.ts:39]: I find the wording 'Review link' a bit too technical. But I also don't really have a better idea for a wording at the moment. If we remove the heading as I suggested in my other comment it could also say "Review links".

File: src/static/data.ts [hunk 3]
```diff
@@ -35,6 +35,7 @@ export const exampleData: Data = [
     points: 1,
     type: QuestionTypes.SingleChoice,
     text: 'What is the capital of France?',
+    referenceLink: 'https://de.wikipedia.org/wiki/Paris',
```
Comment (christophblessing) [inline src/static/data.ts:38]: Since the dataset is in English, can we use links to the English wiki?

File: src/static/data.ts [hunk 4]
```diff
@@ -166,12 +170,216 @@ export const exampleData: Data = [
...
+      {
+        id: 'answer_12',
+        correct: false,
+        text: 'Answer 12',
+      },
+    ],
+  },
+] as Data;
+
+export const testDataWithFourQuestions1: Data = [
```
Comment (christophblessing) [inline src/static/data.ts:273]: This looks unused. Please remove it.

File: src/components/Result.test.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+          remainingAttempts: 0,
+          correctlyAnswered: true,
+        },
+      },
+    ];
+
+    const providerProps = {
+      value: {
+        results,
+        numberOfQuestions: 2,
```
Comment (christophblessing) [inline src/components/Result.test.tsx:36]: It does not seem to make a difference, but there is only one question in the dataset. To avoid confusion I'd say we set it to 1 (same below).

File: src/components/Result.tsx [hunk 6]
```diff
@@ -38,6 +38,11 @@ const Result = () => {
               <th className={`${styles.th} ${styles.centered}`}>
                 {t('result.attempts')}
               </th>
+              {results.some((result) => result.question.referenceLink) ? (
```
Comment (christophblessing) [inline src/components/Result.tsx:41]: I'd prefer the [logical and operator syntax](https://react.dev/learn/conditional-rendering#logical-and-operator-) for the conditional rendering. Safes a few lines and we don't have the empty string in the code.

Comment (christophblessing) [inline src/components/Result.tsx:41]: To improve readability, can you can think about extracting the condition in a variable.

File: src/components/Result.tsx [hunk 7]
```diff
@@ -38,6 +38,11 @@ const Result = () => {
...
+                <th className={styles.th}>{t('result.referenceLink')}</th>
```
Comment (christophblessing) [inline src/components/Result.tsx:42]: For the design we could also omit the table heading text. What do you think?

Screenshot 2024-04-16 at 11 13 52

Don't forget to add an `aria-label` if adapt it like this.

**Review by christophblessing — CHANGES_REQUESTED** [review]
Thanks for opening the PR, nice work!

I added my feedback. Can you please also add a screenshot to the description that show the result of your changes?

File: src/App.test.tsx (hunk 1 above)
Comment (Janis4411) [reply src/App.test.tsx:97]: I agree, not really needed to have this additionally.

File: src/components/Result.tsx (hunk 7 above)
Comment (Janis4411) [reply src/components/Result.tsx:42]: I'd keep the table heading, but if you prefer it without it, I'll remove it.

Comment (christophblessing) [reply src/components/Result.tsx:46]: @JuliaCasamitjana what do you think?

File: src/components/Result.tsx [hunk 8]
```diff
@@ -77,5 +91,3 @@ const Result = () => {
     </div>
   );
 };
-
-export default Result;
```
Comment (christophblessing) [inline src/components/Result.tsx:81]: This doesn't look right - did you remove the export accidentally?

Comment (Janis4411) [reply src/components/Result.tsx:81]: Yep, definitely accidentally.

File: src/App.test.tsx [hunk 9]
```diff
@@ -92,7 +92,6 @@ describe('App component', () => {
       });
 
       const result = screen.queryByText(/Result/);
-
```
Comment (christophblessing) [inline src/App.test.tsx:95]: This change seems unrelated

Comment (Janis4411) [reply src/App.test.tsx:95]: Do you want a line break there? Elsewhere in the file it is also done like that. Declaration and then the expect line.

Comment (christophblessing) [reply src/App.test.tsx:95]: I was wondering because this change is in an unrelated file bloating the diff. But we can keep it.

File: src/components/Result.tsx (hunk 7 above)
Comment (christophblessing) [reply src/components/Result.tsx:46]: Julia and I discussed it in person an we would propose the following UI:

image

The table heading wording is "Reference" and each link in the table is "Link". What do you think?

Comment (Janis4411) [reply src/components/Result.tsx:46]: Yeah I think it is better. I don't think this is perfect yet, but I also don't have a better idea. 😇

**Review by christophblessing — APPROVED** [review]

File: src/components/Result.tsx [hunk 10]
```diff
@@ -56,6 +63,13 @@ const Result = () => {
                   >
                     {`${3 - result.question.remainingAttempts}`}
                   </td>
+                  <td className={styles.td}>
```
Comment (JuliaCasamitjana) [inline src/components/Result.tsx:66]: Does this look center-aligned or should we also use the `centered` class, similar to what we do with the attempts? (same some lines above fot the `th`)

Comment (Janis4411) [reply src/components/Result.tsx:66]: Screenshot 2024-04-17 at 09 47 36

**I'd say it looks better center-aligned as you proposed.** 

Screenshot 2024-04-17 at 09 47 17

Comment (christophblessing) [reply src/components/Result.tsx]: I agree 👍 Please also update the screenshot in the description.

**Review by JuliaCasamitjana — APPROVED** [review]

File: src/components/Result.test.tsx [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      },
+    };
+
+    customRender(<Result />, { providerProps });
+
+    expect(
+      screen.queryByRole('link', { name: 'Reference' }),
+    ).not.toBeInTheDocument();
+    const referenceHeading = screen.queryByRole('columnheader', {
+      name: 'Review link',
```
Comment (christophblessing) [inline src/components/Result.test.tsx:85]: This became a false positive with your changes in the meantime. The test should query for "Reference".

File: src/components/Result.test.tsx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        setQuizEnded: () => {},
+        setQuizStarted: () => {},
+        setResults: () => {},
+      },
+    };
+
+    customRender(<Result />, { providerProps });
+
+    expect(
+      screen.queryByRole('link', { name: 'Reference' }),
```
Comment (christophblessing) [inline src/components/Result.test.tsx:82]: Same here. The text of the link is "Link".
